semantic-ast/docs/json-ASTs.md

```diff
@@ -0,0 +1,42 @@
+
+This document explores how to produce JSON with the hope that it will a) clarify format-specific requirements and b) help us explore the broader design space involving AST outputs across multiple languages and multiple formats. This exercise will likely inform our opinions for the best way to generalize the approach and proceed for other formats (ex., `s-expression` etc).
+
+### Table of Contents
+
+- [Use-cases](#use-cases)
+- [Possible user-facing features/endpoints](#possible-user-facing-features-endpoints)
+- [Design]
+- [Open questions](open-questions)
+
+### Use-cases
+
+Exploring how this JSON would be consumed informs the shape of the output we produce (ex., most likely one JSON object per AST node). Here are a few use-cases that come to mind:
+- External user wants to verify something about an AST, `semantic-ast` gives user an AST that can be piped into `jq` and be used to slice data in various ways
+- External user wants to process the code through _some_ research pipeline
+- GitHub action that parses your code and verifies some property (like a linter)
+- ML team: adjacency list not tree
+- Mobile team: mobile Code Nav features (though this is something to explore using the current pipeline in place, _not_ just precise ASTs)
+- Us! ie., Semantic's program analysis goals
+- TBD
+
+### Possible user-facing features/endpoints
+
+1. **CLI** Currently, Semantic allows us to specify `parse`, `diff`, `graph` along with the `--json` flag to get the output we want. One open question is how closely we want to deviate from this pattern. In `semantic-ast`, we're aiming for a simpler set of options.
+2. **GraphQL** Exposing a GraphQL endpoint different than [the preliminary blob](https://github.com/github/github/blob/master/config/schema.internal.graphql#L9483-L9497) we provide now would be useful for bringing code nav to mobile. Historically we ran into challenges with GraphQL, since it is ideal for data the follows a predetermined structure represented by the schema. Our AST data, however, wouldn't come with this prior knowledge, and the structure would change depending on the user-specified input source code. This wouldn't really allow a user to specify an elaborate query. To get around this, we currently return a single json payload. However, for mobile, we could do something where we figure out how the schema for the data that we get over Twirp RPC in the Aleph query service (not the indexer), and use the structure most viable to the mobile team to inform how to proceed here for `semantic-ast`'s json output.
```

Comment:
Mobile code nav will be relatively straightforward to implement for graphQL and doesn't have the same issues that we ran into with returning tree structures from graphQL. Again, I would vote this is out of scope for semantic-ast.

Reply:
This is absolutely out of scope for `semantic-ast`! I tried clarifying that but I think I should extract that entirely from this design doc. 😄 Would you be game to 🍐 tomorrow on showing me what it would take to get the graphQL endpoint in place? I was poking around Aleph docs today but could use a guide.